src/linked_hash_map.rs

```diff
@@ -0,0 +1,509 @@
+    /// This operation can be used in implemenation of LRU cache.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use collect::LinkedHashMap;
+    /// let mut map = LinkedHashMap::new();
+    ///
+    /// map.insert(1i, "a");
+    /// map.insert(2, "b");
+    /// map.insert(3, "d");
+    ///
+    /// assert_eq!(map.get_refresh(&2), Some(&"b"));
+    ///
+    /// assert_eq!((&2, &"b"), map.iter().rev().next().unwrap());
+    /// ```
+    pub fn get_refresh(&mut self, k: &K) -> Option<&V> {
+        let (value, node_ptr_opt) = match self.map.get_mut(&KeyRef{k: k}) {
+            None => (None, None),
+            Some(node) => {
+                let node_ptr: *mut LinkedHashMapEntry<K, V> = &mut **node;
+                (Some(unsafe { &(*node_ptr).value }), Some(node_ptr))
+            }
+        };
+        match node_ptr_opt {
+            None => (),
+            Some(node_ptr) => {
+                self.detach(node_ptr);
+                self.attach(node_ptr);
+            }
+        }
+        return value;
+    }
+
+    /// Remove and return a value corresponding to the key from the map.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use collect::LinkedHashMap;
+    /// let mut map = LinkedHashMap::new();
+    ///
+    /// map.insert(2i, "a");
+    ///
+    /// assert_eq!(map.remove(&1), None);
+    /// assert_eq!(map.remove(&2), Some("a"));
+    /// assert_eq!(map.remove(&2), None);
+    /// assert_eq!(map.len(), 0);
+    /// ```
+    pub fn remove(&mut self, k: &K) -> Option<V> {
+        let removed = self.map.remove(&KeyRef{k: k});
+        removed.map(|mut node| {
+            let node_ptr: *mut LinkedHashMapEntry<K,V> = &mut *node;
+            self.detach(node_ptr);
+            node.value
+        })
+    }
+
+    /// Return the maximum number of key-value pairs the map can hold without reallocating.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use collect::LinkedHashMap;
+    /// let mut map: LinkedHashMap<int, &str> = LinkedHashMap::new();
+    /// let capacity = map.capacity();
+    /// ```
+    pub fn capacity(&self) -> uint {
+        self.map.capacity()
+    }
+
+    /// Remove the oldest map entry.
+    ///
+    /// Can be used in implementation of LRU cache.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use collect::LinkedHashMap;
+    /// let mut map: LinkedHashMap<int, int> = LinkedHashMap::new();
+    /// map.insert(1, 10);
+    /// map.insert(2, 20);
+    /// map.remove_lru();
+    /// assert_eq!(map.get(&1), None);
+    /// assert_eq!(map.get(&2), Some(&20));
+    /// ```
+    #[inline]
+    pub fn remove_lru(&mut self) {
+        if self.len() > 0 {
+            let lru = unsafe { (*self.head).prev };
+            self.detach(lru);
+            self.map.remove(&KeyRef{k: unsafe { &(*lru).key }});
+        }
+    }
+
+    #[inline]
+    fn detach(&mut self, node: *mut LinkedHashMapEntry<K, V>) {
```

Comment:
Might want to move these to a seperate `impl` block to seperate public/private spatially.